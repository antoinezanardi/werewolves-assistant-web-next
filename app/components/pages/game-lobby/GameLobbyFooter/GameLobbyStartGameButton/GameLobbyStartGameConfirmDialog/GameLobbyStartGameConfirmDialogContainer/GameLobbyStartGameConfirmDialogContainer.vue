<template>
  <div
    id="game-lobby-start-game-confirm-dialog-container"
    class="bg-gray-700 flex flex-col items-center p-4 py-3 rounded"
  >
    <div class="confirm-dialog-header w-full">
      <h5
        class="w-full"
      >
        <i class="fa fa-list-check text-success w-1/12"/>

        <span
          id="confirm-dialog-header-text"
          class="!w-10/12 inline-block text-center"
        >
          {{ $t("components.GameLobbyStartGameConfirmDialog.beforeStartingGame") }}
        </span>
      </h5>

      <PrimeVueDivider class="!my-3"/>
    </div>

    <GameLobbyStartGameConfirmDialogContent
      id="game-lobby-start-game-confirm-dialog-content"
      :current-confirm-step="currentConfirmStep"
      @confirm-step="onConfirmStep"
      @reject-players-position-step="onRejectPlayersPositionStep"
    />

    <GameLobbyStartGameConfirmDialogFooter
      id="game-lobby-start-game-confirm-dialog-footer"
      :current-confirm-step="currentConfirmStep"
      @confirm-start-game="onConfirmStartGame"
      @reject-start-game="onRejectStartGame"
    />
  </div>
</template>

<script setup lang="ts">
import type { GameLobbyStartGameConfirmDialogContainerEmits, GameLobbyStartGameConfirmDialogContainerProps } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/game-lobby-start-game-confirm-dialog-container.types";
import GameLobbyStartGameConfirmDialogContent from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/GameLobbyStartGameConfirmDialogContent/GameLobbyStartGameConfirmDialogContent.vue";
import GameLobbyStartGameConfirmDialogFooter from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/GameLobbyStartGameConfirmDialogFooter/GameLobbyStartGameConfirmDialogFooter.vue";

const props = defineProps<GameLobbyStartGameConfirmDialogContainerProps>();

const emit = defineEmits<GameLobbyStartGameConfirmDialogContainerEmits>();

function onConfirmStep(): void {
  emit("confirmStep");
}

function onRejectPlayersPositionStep(): void {
  emit("rejectPlayersPositionStep", props.rejectCallback);
}

function onConfirmStartGame(): void {
  emit("confirmStartGame");
}

function onRejectStartGame(): void {
  emit("rejectStartGame");
}
</script>