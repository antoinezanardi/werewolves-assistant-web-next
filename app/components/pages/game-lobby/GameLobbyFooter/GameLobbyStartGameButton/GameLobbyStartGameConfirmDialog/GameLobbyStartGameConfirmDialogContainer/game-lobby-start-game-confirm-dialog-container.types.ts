import type { GameLobbyStartGameConfirmDialogStep } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/game-lobby-start-game-confirm-dialog.types";

type GameLobbyStartGameConfirmDialogContainerProps = {
  currentConfirmStep: GameLobbyStartGameConfirmDialogStep;
  rejectCallback: () => void;
};

type GameLobbyStartGameConfirmDialogContainerEmits = {
  confirmStep: [];
  confirmStartGame: [];
  rejectStartGame: [];
  rejectPlayersPositionStep: [rejectCallback: () => void];
  rejectThiefAdditionalCardsPlacedStep: [rejectCallback: () => void];
  rejectActorAdditionalCardsPlacedStep: [rejectCallback: () => void];
  rejectGameOptionsChangedStep: [rejectCallback: () => void];
  rejectPlayersGroupedStep: [rejectCallback: () => void];
};

export type {
  GameLobbyStartGameConfirmDialogContainerProps,
  GameLobbyStartGameConfirmDialogContainerEmits,
};