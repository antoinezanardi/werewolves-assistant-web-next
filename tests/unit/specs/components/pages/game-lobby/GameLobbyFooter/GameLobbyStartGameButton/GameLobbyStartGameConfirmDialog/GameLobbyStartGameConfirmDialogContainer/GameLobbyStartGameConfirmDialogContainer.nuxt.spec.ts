import type { VueVm } from "@tests/unit/utils/types/vue-test-utils.types";
import type { mount } from "@vue/test-utils";

import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type { GameLobbyStartGameConfirmDialogContainerProps } from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/game-lobby-start-game-confirm-dialog-container.types";
import GameLobbyStartGameConfirmDialogContainer from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/GameLobbyStartGameConfirmDialogContainer.vue";
import type GameLobbyStartGameConfirmDialogContent from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/GameLobbyStartGameConfirmDialogContent/GameLobbyStartGameConfirmDialogContent.vue";
import type GameLobbyStartGameConfirmDialogFooter from "~/components/pages/game-lobby/GameLobbyFooter/GameLobbyStartGameButton/GameLobbyStartGameConfirmDialog/GameLobbyStartGameConfirmDialogContainer/GameLobbyStartGameConfirmDialogFooter/GameLobbyStartGameConfirmDialogFooter.vue";

describe("Game Lobby Start Game Confirm Dialog Container Component", () => {
  const defaultProps: GameLobbyStartGameConfirmDialogContainerProps = {
    currentConfirmStep: "players-ready",
    rejectCallback: vi.fn(),
  };
  let wrapper: ReturnType<typeof mount<typeof GameLobbyStartGameConfirmDialogContainer>>;

  async function mountGameLobbyStartGameConfirmDialogContainerComponent(options: ComponentMountingOptions<typeof GameLobbyStartGameConfirmDialogContainer> = {}):
  Promise<ReturnType<typeof mount<typeof GameLobbyStartGameConfirmDialogContainer>>> {
    return mountSuspendedComponent(GameLobbyStartGameConfirmDialogContainer, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    wrapper = await mountGameLobbyStartGameConfirmDialogContainerComponent();
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Header", () => {
    it("should translate header text when rendered.", () => {
      const header = wrapper.find<HTMLHeadingElement>("#confirm-dialog-header-text");

      expect(header.text()).toBe("Before starting the game");
    });
  });

  describe("Dialog Content", () => {
    it("should pass current confirm step as prop when rendered.", () => {
      const dialogContent = wrapper.findComponent<typeof GameLobbyStartGameConfirmDialogContent>("#game-lobby-start-game-confirm-dialog-content");

      expect(dialogContent.props("currentConfirmStep")).toBe("players-ready");
    });

    it("should handle confirm step event when dialog content emits the event.", async() => {
      const dialogContent = wrapper.findComponent<typeof GameLobbyStartGameConfirmDialogContent>("#game-lobby-start-game-confirm-dialog-content");
      (dialogContent.vm as VueVm).$emit("confirmStep");
      await nextTick();

      expect(wrapper.emitted("confirmStep")).toHaveLength(1);
    });

    it("should handle reject players position step event when dialog content emits the event.", async() => {
      const dialogContent = wrapper.findComponent<typeof GameLobbyStartGameConfirmDialogContent>("#game-lobby-start-game-confirm-dialog-content");
      (dialogContent.vm as VueVm).$emit("rejectPlayersPositionStep");
      await nextTick();

      expect(wrapper.emitted("rejectPlayersPositionStep")).toHaveLength(1);
    });
  });

  describe("Dialog Footer", () => {
    it("should pass current confirm step as prop when rendered.", () => {
      const dialogFooter = wrapper.findComponent<typeof GameLobbyStartGameConfirmDialogFooter>("#game-lobby-start-game-confirm-dialog-footer");

      expect(dialogFooter.props("currentConfirmStep")).toBe("players-ready");
    });

    it("should handle confirm start game event when dialog footer emits the event.", async() => {
      const dialogFooter = wrapper.findComponent<typeof GameLobbyStartGameConfirmDialogFooter>("#game-lobby-start-game-confirm-dialog-footer");
      (dialogFooter.vm as VueVm).$emit("confirmStartGame");
      await nextTick();

      expect(wrapper.emitted("confirmStartGame")).toHaveLength(1);
    });

    it("should handle reject start game event when dialog footer emits the event.", async() => {
      const dialogFooter = wrapper.findComponent<typeof GameLobbyStartGameConfirmDialogFooter>("#game-lobby-start-game-confirm-dialog-footer");
      (dialogFooter.vm as VueVm).$emit("rejectStartGame");
      await nextTick();

      expect(wrapper.emitted("rejectStartGame")).toHaveLength(1);
    });
  });
});