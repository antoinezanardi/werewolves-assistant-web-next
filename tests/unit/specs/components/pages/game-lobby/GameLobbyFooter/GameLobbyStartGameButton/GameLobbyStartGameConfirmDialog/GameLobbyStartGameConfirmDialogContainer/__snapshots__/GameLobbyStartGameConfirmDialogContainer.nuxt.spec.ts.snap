// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Start Game Confirm Dialog Container Component > should match snapshot when rendered. 1`] = `
"<div id="game-lobby-start-game-confirm-dialog-container" class="bg-gray-700 flex flex-col items-center p-4 py-3 rounded">
  <div class="confirm-dialog-header w-full">
    <h5 class="w-full">
      <font-awesome-icon-stub icon="list-check" border="false" fixedwidth="false" flip="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" bounce="false" shake="false" beat="false" fade="false" beatfade="false" flash="false" spinpulse="false" spinreverse="false" class="text-success w-1/12"></font-awesome-icon-stub><span id="confirm-dialog-header-text" class="!w-10/12 inline-block text-center">Before starting the game</span>
    </h5>
    <divider-stub class="!my-3"></divider-stub>
  </div>
  <game-lobby-start-game-confirm-dialog-content-stub currentconfirmstep="players-ready" id="game-lobby-start-game-confirm-dialog-content"></game-lobby-start-game-confirm-dialog-content-stub>
  <game-lobby-start-game-confirm-dialog-footer-stub currentconfirmstep="players-ready" id="game-lobby-start-game-confirm-dialog-footer"></game-lobby-start-game-confirm-dialog-footer-stub>
</div>"
`;
